crates/ruff_linter/src/rules/ruff/rules/helpers.rs

```diff
@@ -84,3 +87,72 @@ pub(super) fn is_descriptor_class(func: &Expr, semantic: &SemanticModel) -> bool
         })
     })
 }
+
+// fast check to disqualify any string literal that doesn't
+// have opening and closing brackets
+#[inline]
+fn has_brackets(possible_fstring: &str) -> bool {
+    let mut opening = false;
+    let mut closing = false;
+    for c in possible_fstring.chars() {
+        match c {
+            '{' => opening = true,
+            '}' => closing = true,
+            _ => {}
+        }
+    }
+    opening && closing
+}
+
+/// Returns `true`` if `source` is valid f-string syntax with qualified, bound variables.
+/// `kwargs` should be the keyword arguments that were passed to function if the string literal is also
+/// being passed to the same function.
+/// If a identifier from `kwargs` is used in `source`'s formatting, this will return `false`,
+/// since it's possible the function could be formatting the literal in question.
+pub(super) fn should_be_fstring(
+    source: &str,
+    kwargs: Option<&Vec<Keyword>>,
+    semantic: &SemanticModel,
+) -> bool {
+    if !has_brackets(source) {
+        return false;
+    }
+
+    let Ok(Expr::FString(ast::ExprFString { value, .. })) =
+        parse_expression(&format!("f\"{source}\""))
```

Comment:
This needs to be changed to accept the original string literal using `locator.slice(...)`

Reply:
Can you explain why you would change it to use `locator.slice`? I fear that you could run into issue for parenthesized multline string literals.

Reply:
Can we avoid adding the quotes here and instead just add the prefix with the original string? I think that would require `locator.slice(...)` to extract to literal.

> I fear that you could run into issue for parenthesized multline string literals.

I'm not sure if that would be a problem here because, per my understanding, this function is called for individual string literal and not the concatenated one.

Reply:
> I'm not sure if that would be a problem here because, per my understanding, this function is called for individual string literal and not the concatenated one.

Right, this is for individual string literals.